fms/models/llama.py

```diff
@@ -573,9 +593,120 @@ def _hf_sd_to_fms_sd(hf_sd: Mapping) -> Mapping:
+            if temp.size(0) == 2560:
+                head_size = 80  # granite 3b code
+            else:
+                head_size = 128  # every other Llama model in existence
+            nheads = int(temp.size(0) / head_size)
+
+            if temp.dim() == 2:  # weight
+                temp_view = temp.view(nheads, 2, -1, temp.size(1))
+            else:  # bias
+                temp_view = temp.view(nheads, 2, -1)
+            temp = temp_view.transpose(1, 2).reshape(*temp.size())
+
+            new_sd[new_name] = temp
+
+    return new_sd
+
+
+# TODO: merge with standard _hf_sd_to_fms_sd adapter
+# Very similar except:
+# 1) add in_proj to q,k,v
+# 2) qparams to transpose are qweight|scales|qzeros (not weight|bias)
+# 3) fully transpose params before & after processing
+def _gptq_unfused_sd_to_fms_unfused_sd(hf_sd: Mapping) -> Mapping:
+    replacements = [
+        (r"^lm_head.weight", "shared.head.weight"),
+        (r"^model.embed_tokens.weight", "shared.emb.weight"),
+        (r"^model.norm", "dec_norm"),
+        (r"^model.layers", "layers"),
+        (r"self_attn\.k_proj", "attn.in_proj.key"),
+        (r"self_attn\.v_proj", "attn.in_proj.value"),
+        (r"self_attn\.q_proj", "attn.in_proj.query"),
+        (r"self_attn\.o_proj", "attn.dense"),
+        (r"mlp\.gate_proj", "ff_sub_layer.wg"),
+        (r"mlp\.up_proj", "ff_sub_layer.w1"),
+        (r"mlp\.down_proj", "ff_sub_layer.w2"),
+        (r"input_layernorm", "ln"),
+        (r"post_attention_layernorm", "ff_ln"),
+    ]
+    fms_unfused_sd = {}
+    trans_required_pattern = re.compile(
+        "layers.[0-9]+.attn.in_proj.(query|key).(qweight|scales|qzeros)"
+    )
+    for hf_name, param in hf_sd.items():
+        fms_name = hf_name
+        for pattern, repl in replacements:
+            fms_name = re.sub(pattern, repl, fms_name)
+        fms_unfused_sd[fms_name] = param
+
+        # hf -> fms requires a transpose operation for the query and key
+        # weight and bias parameters
+        # Differently from hf-to-fms conversion, GPTQ qweights are [in_feat, out_feat]
+        # and are fully transposed before & after process
+        # FIXME: is unpack/transpose/repack for qweight and qzeros also needed?
+        # is there any combination of group quantization that requires unpacking?
+        if bool(trans_required_pattern.match(fms_name)):
+            param_t_size = param.t().size()
+
+            # TODO: less hardcoded way to determine head_size?
+            if param_t_size[1] == 2560:
+                head_size = 80  # granite 3b code
+            else:
+                head_size = 128  # every other Llama model in existence
+            nheads = int(param_t_size[0] / head_size)
+
+            fms_unfused_sd[fms_name] = (
+                param.t().view(nheads, 2, -1, param_t_size[1])
+                .transpose(1, 2)
+                .reshape(*param_t_size).t()
+            )
+    return fms_unfused_sd
```

Comment:
the way to merge all these is the following:

```python
def _hf_sd_to_fms_sd(hf_sd: Mapping, fuse: bool = True) -> Mapping:
    replacements = [
        (r"^lm_head.weight", "shared.head.weight"),
        (r"^model.embed_tokens.weight", "shared.emb.weight"),
        (r"^model.norm", "dec_norm"),
        (r"^model.layers", "layers"),
        (r"self_attn\.k_proj", "attn.in_proj.key"),
        (r"self_attn\.v_proj", "attn.in_proj.value"),
        (r"self_attn\.q_proj", "attn.in_proj.query"),
        (r"self_attn\.o_proj", "attn.dense"),
        (r"mlp\.gate_proj", "ff_sub_layer.wg"),
        (r"mlp\.up_proj", "ff_sub_layer.w1"),
        (r"mlp\.down_proj", "ff_sub_layer.w2"),
        (r"input_layernorm", "ln"),
        (r"post_attention_layernorm", "ff_ln"),
    ]
    new_sd = {}
    trans_required_pattern = re.compile("layers.[0-9]+.attn.in_proj.(query|key).(weight|bias|qweight|scales|qzeros)")
    for name, param in hf_sd.items():
        new_name = name
        for pattern, repl in replacements:
            new_name = re.sub(pattern, repl, new_name)
        new_sd[new_name] = param
        # hf -> fms requires a transpose operation for the query and key
        # weight and bias parameters
        # This transpose is due to the different implementation of RoPE in
        # HF and FMS. While FMS follows the original RoPE paper
        # (https://arxiv.org/abs/2104.09864), HF has its own implementation
        # that doesn't respect the order of outputs. This is OK as long as you
        # rearrange the weights of the query and key projections, as the
        # combination projection + RoPE ends up producing the same outputs.
        # Therefore, to make FMS produce the correct order of outputs when
        # loading from an HF checkpoint, we need to undo the transformation
        # that HF does from the original Meta weights:
        if bool(trans_required_pattern.match(new_name)):
     
            temp = new_sd[new_name]
            # nheads is used in the transformation required for hf->fms
            if temp.size(0) == 2560:
                head_size = 80  # granite 3b code
            else:
                head_size = 128  # every other Llama model in existence
            nheads = int(temp.size(0) / head_size)
            if temp.dim() == 2:  # weight
                temp_view = temp.view(nheads, 2, -1, temp.size(1))
            else:  # bias
                temp_view = temp.view(nheads, 2, -1)
            temp = temp_view.transpose(1, 2).reshape(*temp.size())
            new_sd[new_name] = temp
    if fuse:
        return _convert_to_fused(new_sd)
    return fused_sd
```

Reply:
I think there's a problem in this proposed merging of the original adapter with the GPTQ one: the original `_hf_sd_to_fms_sd` converts `self_attn.k_proj` to `attn.query` (legacy FMS unfused format), not to `attn.in_proj.query` (current FMS unfused format). Then, within `_convert_to_fused`, `attn.query` is converted to `attn.in_proj.qkv_fused` but `attn.in_proj.query` would not be handled.

Reply:
we can change this to use `in_proj` everywhere, as it was probably an oversight of the first fused work

Reply:
@JRosenkranz anything I'm missing?

Reply:
I think that should be fine, we should always have in_proj. Only case we don't is with legacy fms weights

Reply:
can be left as a TODO for the FP8 rework